### PR TITLE
Implement OneHot CUDA Kernels

### PR DESCRIPTION
**Pull request by SherlockNoMad**
**Description**: Implement OneHot CUDA Kernels

**Motivation and Context**
- Needed for UniLM v2 Model Training

### CODE REVIEW

File: onnxruntime/core/providers/cpu/tensor/onehot.cc [hunk 1]
```diff
@@ -85,6 +84,39 @@ Status ValidateInputs(const Tensor* depth,
   return Status::OK();
 }
 
+Status PrepareOutputShape(const Tensor* indices, const int64_t depth_val, const int64_t axis,
+                          int64_t& prefix_dim_size, int64_t& suffix_dim_size,
+                          std::vector<int64_t>& output_shape) {
+  const auto& indices_shape = indices->Shape();
+  const auto& indices_dims = indices_shape.GetDims();
+  const auto indices_num_dims = indices_shape.NumDimensions();
+  //std::vector<int64_t> output_shape(indices_shape.GetDims());
```
Comment (pengwa) [inline onnxruntime/core/providers/cpu/tensor/onehot.cc:93]: remove this.

File: onnxruntime/test/providers/cpu/tensor/onehot_op_test.cc [hunk 2]
```diff
@@ -40,7 +40,7 @@ TEST(OneHotOpTest, DefaultAxis_float_float_float /*indices, output, depth*/) {
 }
 
 TEST(OneHotOpTest, DefaultAxis_int64_int32_float /*indices, output, depth*/) {
-  OpTester test("OneHot", 9);
+  OpTester test("OneHot", 11);
```
Comment (pengwa) [inline onnxruntime/test/providers/cpu/tensor/onehot_op_test.cc:43]: so we don't want to keep opset 9 tests? I know opset 11 OneHot is a superset of opset 9 though. One possible thing I thought is: a kind of cpu provider implement OPSET 9 using a way different with 11. Your call for this. :)

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 3]
```diff
@@ -705,6 +705,9 @@ class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain,
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 10, int8_t, DequantizeLinear);
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 10, uint8_t, DequantizeLinear);
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 11, CumSum);
+class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 11, int64_t_int64_t_int64_t, OneHot);
```
Comment (pengwa) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc:708]: do we need add fp16 support for the output?

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 4]
```diff
@@ -727,7 +730,6 @@ class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain,
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 12, int8_t, ReduceMin);
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCudaExecutionProvider, kOnnxDomain, 12, uint8_t, ReduceMin);
 
-
 static void RegisterCudaKernels(KernelRegistry& kernel_registry) {
```
Comment (pengwa) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc:733]: nit: deleting this line by intention?

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc [hunk 5]
```diff
@@ -1368,12 +1372,12 @@ CUDAExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph,
     for (auto registry : kernel_registries) {
       cuda_kernel_def = registry->TryFindKernel(node, Type());
 
-      // atleast one registry has a CUDA kernel for this node 
+      // atleast one registry has a CUDA kernel for this node
```
Comment (pengwa) [inline onnxruntime/core/providers/cuda/cuda_execution_provider.cc:1375]: at least

File: onnxruntime/core/providers/cuda/tensor/onehot.cu [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  int prefix_index, prefix_offset;
+  fdm_depth_suffix.divmod(id, prefix_index, prefix_offset);
+
+  int depth_index, suffix_index;
+  fdm_suffix.divmod(prefix_offset, depth_index, suffix_index);
+
+  CUDA_LONG indices_index = prefix_index * fdm_suffix.d_ + suffix_index;
+
+  // handle negative index
+  in_type adjusted_indice = (indices_data[indices_index] + depth_val) % depth_val;
```
Comment (pengwa) [inline onnxruntime/core/providers/cuda/tensor/onehot.cu:32]: I don't know how % performs on gpu. do you think putting this in onehot.cc be better (just as CPU OneHot kernel did).

File: onnxruntime/core/providers/cuda/tensor/onehot.cu [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+  output_data[id] = (adjusted_indice == in_type(depth_index)) ? on_value : off_value;
+}
```
Comment (pengwa) [inline onnxruntime/core/providers/cuda/tensor/onehot.cu:35]: just confirm, is this unconmmon case covered? I guess not?

"Any entries in the 'indices' input tensor with values outside the range [-depth, depth-1] will result in one-hot representation with all 'off_value' values in the output tensor.

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc (hunk 4 above)
Comment (SherlockNoMad) [reply onnxruntime/core/providers/cuda/cuda_execution_provider.cc:733]: Yes. Auto formated

File: onnxruntime/core/providers/cuda/cuda_execution_provider.cc (hunk 3 above)
Comment (SherlockNoMad) [reply onnxruntime/core/providers/cuda/cuda_execution_provider.cc:708]: yes. added

File: onnxruntime/test/providers/cpu/tensor/onehot_op_test.cc (hunk 2 above)
Comment (SherlockNoMad) [reply onnxruntime/test/providers/cpu/tensor/onehot_op_test.cc:43]: opset 11 is a superset of opset 9. 
and they share the same code path. 
I think updating the test opset set to 11 is enough to cover the kernel imple.

File: onnxruntime/core/providers/cuda/tensor/onehot.cu (hunk 6 above)
Comment (SherlockNoMad) [reply onnxruntime/core/providers/cuda/tensor/onehot.cu:32]: indices_data lives in GPU memory. 
I think we have no choice but to handle the negative index in cuda kernel.

File: onnxruntime/core/providers/cuda/tensor/onehot.cu (hunk 7 above)
Comment (SherlockNoMad) [reply onnxruntime/core/providers/cuda/tensor/onehot.cu:38]: Good catch.  Fixed.

File: onnxruntime/core/providers/cuda/tensor/onehot.cu (hunk 6 above)
Comment (pengwa) [reply onnxruntime/core/providers/cuda/tensor/onehot.cu:35]: you are right!

File: onnxruntime/core/providers/cpu/tensor/onehot.cc [hunk 8]
```diff
@@ -85,6 +84,38 @@ Status ValidateInputs(const Tensor* depth,
...
+  if (axis >= output_rank || axis < -output_rank) {
+    std::ostringstream oss;
+    oss << "'axis' attribute must have a value in the range [" << -output_rank
+        << "," << indices_num_dims << "]";
+    return Status(ONNXRUNTIME, INVALID_ARGUMENT, oss.str());
+  }
+
+  auto true_axis = axis;
+  if (true_axis < 0)
+    true_axis += output_rank;
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cpu/tensor/onehot.cc:106]: there's a util func for HandleNegativeAxis

File: onnxruntime/core/providers/cpu/tensor/onehot.cc [hunk 9]
```diff
@@ -85,6 +84,38 @@ Status ValidateInputs(const Tensor* depth,
...
+
+  output_shape.insert(output_shape.begin() + true_axis, depth_val);
+
+  prefix_dim_size = 1;
+  for (int64_t i = 0; i < true_axis; ++i) {
+    prefix_dim_size *= indices_dims[i];
```
Comment (KeDengMS) [inline onnxruntime/core/providers/cpu/tensor/onehot.cc:112]: >indices_dims [](start = 23, length = 12)

could this be scalar? Then indices_dims is 0-size

Comment (SherlockNoMad) [reply onnxruntime/core/providers/cpu/tensor/onehot.cc:112]: No. Spec Shape Inference enforces this 
" Indices tensor must have rank >= 1"

**Pull request merged**